models/demos/wormhole/distilbert/tt/ttnn_optimized_distilbert.py

```diff
@@ -0,0 +1,419 @@
+        parameters.query_key_value.weight,
+        bias=parameters.query_key_value.bias,
+        memory_config=ttnn.L1_MEMORY_CONFIG,
+        dtype=ttnn.bfloat8_b,
+        core_grid=ttnn.CoreGrid(y=batch_size, x=num_cores_x),
+    )
+
+    (
+        query,
+        key,
+        value,
+    ) = ttnn.transformer.split_query_key_value_and_split_heads(
+        query_key_value_output,
+        memory_config=ttnn.L1_MEMORY_CONFIG,
+        num_heads=config.n_heads,
+    )
+    ttnn.deallocate(query_key_value_output)
+
+    query = query * (1 / (dim_per_head) ** 0.5)
+
+    attention_scores = ttnn.matmul(
+        query,
+        key,
+        memory_config=ttnn.L1_MEMORY_CONFIG,
+        dtype=ttnn.bfloat16,
+        core_grid=ttnn.CoreGrid(y=batch_size, x=num_cores_x),
+    )
+    ttnn.deallocate(query)
+    ttnn.deallocate(key)
+    score_list = []
+
+    if batch_size <= 2:
+        inter_scores = attention_scores * negative_val_tensor
+        inter_scores = inter_scores + attention_scores
+        scores = inter_scores + min_val_tensor
+    else:
+        for i in range(2, batch_size + 1, 2):
+            inter_scores = attention_scores[i - 2 : i, :, :, :] * negative_val_tensor[i - 2 : i, :, :, :]
+            inter_scores = inter_scores + attention_scores[i - 2 : i, :, :, :]
+            score = inter_scores + min_val_tensor[i - 2 : i, :, :, :]
+            score = ttnn.permute(score, (1, 0, 2, 3))
+
+            score_list.append(score)
+            ttnn.deallocate(inter_scores)
+
+        scores = ttnn.concat(score_list, dim=1)
+        scores = ttnn.permute(scores, (1, 0, 2, 3))
+
+    weights = ttnn.transformer.attention_softmax(scores, head_size=1)
+    ttnn.deallocate(scores)
+    context_layer = ttnn.matmul(
+        weights,
+        value,
+        memory_config=ttnn.L1_MEMORY_CONFIG,
+        dtype=ttnn.bfloat16,
+        core_grid=ttnn.CoreGrid(y=batch_size, x=num_cores_x),
+    )
+
+    ttnn.deallocate(weights)
+    ttnn.deallocate(value)
+
+    context_layer = ttnn.permute(context_layer, [0, 1, 3, 2])
+    context_layer = ttnn.reshape(context_layer, (batch_size, config.n_heads * dim_per_head, -1))
+
+    context_layer = ttnn.permute(context_layer, (0, 2, 1))
+
+    self_output = ttnn.linear(
+        context_layer,
+        parameters.out_lin.weight,
+        bias=parameters.out_lin.bias,
+        memory_config=ttnn.DRAM_MEMORY_CONFIG,
+        dtype=ttnn.bfloat16,
+        core_grid=ttnn.CoreGrid(y=batch_size, x=num_cores_x),
+    )
+    ttnn.deallocate(context_layer)
+    return self_output
+
+
+def ffn(configs, hidden_state, device, base_address, parameters, num_cores_x=12, mesh_mapper=None):
+    batch_size, *_ = hidden_state.shape
+
+    output = ttnn.linear(
+        hidden_state,
+        parameters.lin1.weight,
+        bias=parameters.lin1.bias,
+        memory_config=ttnn.L1_MEMORY_CONFIG,
+        dtype=ttnn.bfloat16,
+        activation="gelu",
+        core_grid=ttnn.CoreGrid(y=batch_size, x=num_cores_x),
```

Comment:
I don't understand using the batch size for num_cores_y. Why not just use all the cores and let ttnn.linear decide how to use them? The rows will be used for batch_size * seq_len elements, so if seq_len is >32 we can use more than 1 core per batch element for example.

Reply:
@yieldthought ,
the `core_grid` is replaced with `core_grid=ttnn.CoreGrid(y=device.core_grid.y, x=device.core_grid.x)`.
I hope that this approach will manage the distribution effectively, utilizing all available cores while allowing ttnn.linear to determine their optimal usage.